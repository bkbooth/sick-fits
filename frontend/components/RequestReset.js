import React, { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import Error from 'components/ErrorMessage';
import Form from 'components/styles/Form';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

const RequestReset = () => {
  const [email, setEmail] = useState('');
  const [requestReset, { called, error, loading }] = useMutation(REQUEST_RESET_MUTATION, {
    variables: { email },
  });

  function createChangeHandler(setter) {
    return event => setter(event.target.value);
  }

  async function handleRequestReset(event) {
    event.preventDefault();
    await requestReset();
    setEmail('');
  }

  return (
    <Form onSubmit={handleRequestReset} method="post" data-test="form">
      <fieldset disabled={loading} aria-busy={loading}>
        <h2>Request a password reset</h2>

        <Error error={error} />

        {!error && !loading && called && (
          <p data-test="success">Your password reset link has been sent to your email address.</p>
        )}

        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            id="email"
            placeholder="Email"
            value={email}
            onChange={createChangeHandler(setEmail)}
            required
          />
        </label>

        <button type="submit">Request{loading ? 'ing' : ''} reset</button>
      </fieldset>
    </Form>
  );
};

export default RequestReset;
