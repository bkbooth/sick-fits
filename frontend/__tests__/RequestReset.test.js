import { MockedProvider } from '@apollo/react-testing';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import RequestReset, { REQUEST_RESET_MUTATION } from 'components/RequestReset';

const mocks = [
  {
    request: { query: REQUEST_RESET_MUTATION, variables: { email: 'test@example.com' } },
    result: { data: { requestReset: { __typename: 'Message', message: 'Success!' } } },
  },
];

describe('<RequestReset />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(wrapper.find('form[data-test="form"]')).toMatchSnapshot();
  });

  it('calls the request reset mutation', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: 'test@example.com' } });
    wrapper.find('form').simulate('submit');
    await act(() => wait());
    wrapper.update();
    expect(wrapper.find('p[data-test="success"]').text()).toContain(
      'Your password reset link has been sent to your email address.'
    );
  });
});
