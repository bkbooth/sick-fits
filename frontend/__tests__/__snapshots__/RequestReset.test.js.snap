// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Request a password reset
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Request
       reset
    </button>
  </fieldset>
</form>
`;
